django/queryset/README.md

```diff
@@ -0,0 +1,340 @@
+
+
+class Person(models.Model):
+    user = models.ForeignKey(
+        'User',
+        on_delete=models.CASCADE,
+        related_name='people',
+    )
+    first_name = models.CharField(max_length=50)
+    last_name = models.CharField(max_length=150)
+    employee_id = models.CharField(max_length=15, unique=True)
+    active = models.Boolean(default=True, blank=True)
+
+
+class Department(models.Model):
+    name = models.CharField(max_length=255)
+
+    members = models.ManyToManyField(
+        'Person',
+        related_name='departments',
+    )
+```
+
+## Filtering
+
+### Joint Field Filtering
+
+```python
+# Yes
+def find_people_yes(query: str):
+    return Person.objects.annotate(
+        full_name=Concat('first_name', Value(' '), 'last_name'),
+    ).filter(
+        full_name__icontains=query,
+    )
+
+# No
+def find_people_no(query: str):
+    return Person.objects.filter(
+        Q(first_name__icontains=query)
+        | Q(last_name__icontains=query)
+    )
+```
+
+First, we try to filter people with name `John`...
+
+```python
+people = find_people_yes(query='John')
+# return: <QuerySet [<Person: John Doe>, <Person: John Wick>]>
+
+people = find_people_no(query='John')
+# return: <QuerySet [<Person: John Doe>, <Person: John Wick>]>
+```
+
+Everything seems fine, right? Both techniques return the same results. But what about filtering person's name by `John Doe`?
+
+```python
+people = find_people_yes(query='John Doe')
+# return: <QuerySet [<Person: John Doe>]>
+
+people = find_people_no(query='John Doe')
+# return: <QuerySet []>
+```
+
+Now you see the difference? It's because the second method (`find_people_no`) does not support filtering across two fields.
+
+So if you have a requirement like this, you should join two fields together before filter them.
+
+### Using `.count()` instead of `len()`
+
+```python
+# Yes
+Person.objects.count()     # SELECT COUNT(*) FROM person
+
+# No
+len(Person.objects.all())  # SELECT * FROM person
+```
+
+Selecting aggregate function from database always consume less time that getting every records.
+
+### Using `.exists()` to check if any data exists
+
+```python
+# Yes
+if Person.objects.filter(first_name='John').exists():
+    do_something()
+
+# No
+if Person.objects.filter(first_name='John'):
+    do_something()
+```
+
+Using `.exists()` consume less resources than not using it.
+
+### Using `.first()` instead of calling `.exists()`, followed by `.get()`
+
+```python
+# Yes
+matched_person = Person.objects.filter(employee_id='123').first()
+if matched_person is not None:
+    do_something(matched_person)
+
+# No
+if Person.objects.filter(employee_id='123').exists():
+    matched_person = Person.objects.get(employee_id='123')
+    do_something(matched_person)
+```
+
+The reason is calling a single `.first()` will hit database only one. But when we call `.exists()`, it hits. Calling `.get()` will hits database again. Makes the second method to hit database twice which is a bad idea.
```

Comment:
I got confuse after reading this. I'm suggesting this instead: 
```
Calling `.first()` method will hit database only one time, and you will also get the object.
Calling `.exists()` method hit database one time, then calling `.get()` hit database one more time.
In this case, using `.first()` has better performance since it hit database only one.
```

Reply:
Done krub. Also added `.only()`. But about query evaluation I only put the reference to Django documentation na krub because I think their documentation explained it very well and it is not something to define in a standard. In my opinion, it is something everyone should know 55555555.

Reply:
Agree krub